packages/core/react/src/WalletProvider.tsx

```diff
@@ -23,285 +24,158 @@ export interface WalletProviderProps {
+            return null;
         }
-
-        adapters.forEach((adapter) => adapter.on('readyStateChange', handleReadyStateChange, adapter));
-        return () => adapters.forEach((adapter) => adapter.off('readyStateChange', handleReadyStateChange, adapter));
-    }, [adapters]);
-
-    // When the selected wallet changes, initialize the state
-    useEffect(() => {
-        const wallet = name && wallets.find(({ adapter }) => adapter.name === name);
-        if (wallet) {
-            setState({
-                wallet,
-                adapter: wallet.adapter,
-                connected: wallet.adapter.connected,
-                publicKey: wallet.adapter.publicKey,
-            });
-        } else {
-            setState(initialState);
+        const existingMobileWalletAdapter = adapters.find(
+            (adapter) => adapter.name === SolanaMobileWalletAdapterWalletName
+        );
+        if (existingMobileWalletAdapter) {
+            return existingMobileWalletAdapter;
         }
-    }, [name, wallets]);
-
-    // If the window is closing or reloading, ignore disconnect and error events from the adapter
-    useEffect(() => {
-        function listener() {
-            isUnloading.current = true;
+        return new SolanaMobileWalletAdapter({
+            addressSelector: createDefaultAddressSelector(),
+            appIdentity: {
+                uri: getUriForAppIdentity(),
+            },
+            authorizationResultCache: createDefaultAuthorizationResultCache(),
+            cluster: getInferredClusterFromEndpoint(connection?.rpcEndpoint),
+        });
+    }, [adapters, connection?.rpcEndpoint]);
+    const adaptersWithDefaultsInjected = useMemo(() => {
+        if (!getIsMobile(adapters)) {
+            return adapters;
         }
-
-        window.addEventListener('beforeunload', listener);
-        return () => window.removeEventListener('beforeunload', listener);
-    }, [isUnloading]);
-
-    // Handle the adapter's connect event
-    const handleConnect = useCallback(() => {
-        if (!adapter) return;
-        setState((state) => ({ ...state, connected: adapter.connected, publicKey: adapter.publicKey }));
-    }, [adapter]);
-
-    // Handle the adapter's disconnect event
-    const handleDisconnect = useCallback(() => {
-        // Clear the selected wallet unless the window is unloading
-        if (!isUnloading.current) setName(null);
-    }, [isUnloading, setName]);
-
-    // Handle the adapter's error event, and local errors
-    const handleError = useCallback(
-        (error: WalletError) => {
-            // Call onError unless the window is unloading
-            if (!isUnloading.current) (onError || console.error)(error);
-            return error;
-        },
-        [isUnloading, onError]
+        if (adapters.indexOf(mobileWalletAdapter as Adapter) !== -1) {
+            return adapters;
+        }
+        return [mobileWalletAdapter, ...adapters] as Adapter[];
```

Comment:
nit: narrowly cast only MWA since we know it's not `null` from `getIsMobile`
```suggestion
        if (adapters.indexOf(mobileWalletAdapter as Adapter) !== -1) {
            return adapters;
        }
        return [mobileWalletAdapter as Adapter, ...adapters];
```